controller/appcontroller.go

```diff
@@ -1874,7 +1874,8 @@ func (ctrl *ApplicationController) autoSync(app *appv1.Application, syncStatus *
 	// auto-sync with pruning disabled). We need to ensure that we do not keep Syncing an
 	// application in an infinite loop. To detect this, we only attempt the Sync if the revision
 	// and parameter overrides are different from our most recent sync operation.
-	if alreadyAttempted && (!selfHeal || !attemptPhase.Successful()) {
+	// Unless selfHeal if set to true in which case we will retry
+	if alreadyAttempted && !selfHeal {
```

Comment:
If think the if logic could be made more readable. 
```
if alreadyAttempted {
  if attemptPhase.Successful() {
     // log "Skipping auto-sync: most recent sync already to %s"
     return
  }
  if !selfheal {
    // log "Skipping auto-sync: failed previous sync attempt to %s"
    return
  }
  if !shouldSelfheal {
    //log "Skipping auto-sync: already attempted sync to %s with timeout %v (retrying in %v)"
    return
  }
  // do self heal
}
```

Reply:
Maybe, but perhaps the immediate focus should be fixing the rather obvious bug.